### PR TITLE
[MOVED] Add support for extended precision integers and decimals

### PR DESCRIPTION
**Pull request by andresokol**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

...

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

File: dbms/src/Core/Types.h [hunk 1]
```diff
@@ -240,4 +261,14 @@ namespace std
                 ^ std::hash<DB::Int64>()(x.value & std::numeric_limits<DB::UInt64>::max());
         }
     };
+
+    template <>
+    struct hash<DB::Decimal256>
+    {
+        size_t operator()(const DB::Decimal256 & x) const
+        {
+            return std::hash<DB::Int128>()(x.value >> 128)
```
Comment (alexey-milovidov) [inline dbms/src/Core/Types.h]: This is a classic mistake.

Please watch this video: https://www.youtube.com/watch?v=EoX82TEz2sQ
https://presentations.clickhouse.tech/hash_tables/#11

File: dbms/src/Core/Types.h [hunk 2]
```diff
@@ -55,11 +59,13 @@ using UInt8 = uint8_t;
 using UInt16 = uint16_t;
 using UInt32 = uint32_t;
 using UInt64 = uint64_t;
+using UInt256 = boost::multiprecision::uint256_t;
```
Comment (qoega) [inline dbms/src/Core/Types.h]: What about UInt128/Int128?

File: dbms/src/DataTypes/DataTypesNumber.h [hunk 3]
```diff
@@ -38,10 +38,12 @@ using DataTypeUInt8 = DataTypeNumber<UInt8>;
 using DataTypeUInt16 = DataTypeNumber<UInt16>;
 using DataTypeUInt32 = DataTypeNumber<UInt32>;
 using DataTypeUInt64 = DataTypeNumber<UInt64>;
+using DateTypeUInt256 = DataTypeNumber<UInt256>;
```
Comment (qoega) [inline dbms/src/DataTypes/DataTypesNumber.h]: DataTypeUInt128/DataTypeInt128 are lost

**Comment by 4ertus2** [comment]
You also need implement `*Overflow()` functions for 256-bit types here https://github.com/ClickHouse/ClickHouse/blob/master/base/common/arithmeticOverflow.h#L30

**Comment by alesapin** [comment]
Do not test until merge with master.

**Comment by alexey-milovidov** [comment]
Continued in #10388

**Pull request closed without merging**